pkg/dbmate/db.go

```diff
@@ -324,51 +324,49 @@ func (db *DB) Migrate() error {
 		}
 	}
 
-	if len(pendingMigrations) == 0 {
-		return nil
-	}
-
-	if db.Strict && pendingMigrations[0].Version <= highestAppliedMigrationVersion {
-		return fmt.Errorf("migration `%s` is out of order with already applied migrations, the version number has to be higher than the applied migration `%s` in --strict mode", pendingMigrations[0].Version, highestAppliedMigrationVersion)
-	}
-
-	sqlDB, err := db.openDatabaseForMigration(drv)
-	if err != nil {
-		return err
-	}
-	defer dbutil.MustClose(sqlDB)
-
-	for _, migration := range pendingMigrations {
-		fmt.Fprintf(db.Log, "Applying: %s\n", migration.FileName)
+	if len(pendingMigrations) > 0 {
+		if db.Strict && pendingMigrations[0].Version <= highestAppliedMigrationVersion {
+			return fmt.Errorf("migration `%s` is out of order with already applied migrations, the version number has to be higher than the applied migration `%s` in --strict mode", pendingMigrations[0].Version, highestAppliedMigrationVersion)
+		}
 
-		parsed, err := migration.Parse()
+		sqlDB, err := db.openDatabaseForMigration(drv)
 		if err != nil {
 			return err
 		}
+		defer dbutil.MustClose(sqlDB)
 
-		execMigration := func(tx dbutil.Transaction) error {
-			// run actual migration
-			result, err := tx.Exec(parsed.Up)
+		for _, migration := range pendingMigrations {
+			fmt.Fprintf(db.Log, "Applying: %s\n", migration.FileName)
+
+			parsed, err := migration.Parse()
 			if err != nil {
 				return err
-			} else if db.Verbose {
-				db.printVerbose(result)
 			}
 
-			// record migration
-			return drv.InsertMigration(tx, migration.Version)
-		}
+			execMigration := func(tx dbutil.Transaction) error {
+				// run actual migration
+				result, err := tx.Exec(parsed.Up)
+				if err != nil {
+					return err
+				} else if db.Verbose {
+					db.printVerbose(result)
+				}
+
+				// record migration
+				return drv.InsertMigration(tx, migration.Version)
+			}
 
-		if parsed.UpOptions.Transaction() {
-			// begin transaction
-			err = doTransaction(sqlDB, execMigration)
-		} else {
-			// run outside of transaction
-			err = execMigration(sqlDB)
-		}
+			if parsed.UpOptions.Transaction() {
+				// begin transaction
+				err = doTransaction(sqlDB, execMigration)
+			} else {
+				// run outside of transaction
+				err = execMigration(sqlDB)
+			}
 
-		if err != nil {
-			return err
+			if err != nil {
+				return err
+			}
 		}
 	}
```

Comment:
Instead of including all of this inside the outer `if` statement starting on line 327, would just using this work, instead:

```
	if len(pendingMigrations) > 0 && db.Strict && pendingMigrations[0].Version <= highestAppliedMigrationVersion {
		return fmt.Errorf("migration `%s` is out of order with already applied migrations, the version number has to be higher than the applied migration `%s` in --strict mode", pendingMigrations[0].Version, highestAppliedMigrationVersion)
	}
```

This reduces the amount of code executed inside the conditional, and achieves the same effect, I believe.  What do you think?

Reply:
It would work yes, as the condition is only really necessary for that part. But everything afterwards within the if statement benefits too, right? We do not need to open the database if we do not do any migrations. But I am not sure how multiple calls to `drv.Open()` are handled or if they are optimized by go either way. The for loop will most definitely be optimized though and does not need to be part of the if statement.

Reply:
It's not a matter of multiple calls to `drv.Open()` but the fact that moving the call to `db.openDatabaseForMigration()` inside the `if` condition means that any side effects of `db.openDatabaseForMigration()` (such as creating, or at least attempting to create, the migrations table) also won't happen.

The original code, when there are no migrations to apply, will still create, or attempt to create, the migrations table.

With your code as it is currently, that won't happen.  This is a subtle but meaningful difference in behavior, that if someone else is depending on, this change will break on them.

Unless there's a good reason to introduce a backwards-incompatible breaking change, I would prefer not to.

Reply:
Right, I totally missed that `db.FindMigrations()` only checks if the migration table exists but does not create the migration tables like `db.openDatabaseForMigration()`. I fixed the if statement. Thanks for pointing that out 👍